___

<a href='http://www.pieriandata.com'> <img src='../../Pierian_Data_Logo.png' /></a>
___

# SF Salaries Exercise 

Welcome to a quick exercise for you to practice your pandas skills! We will be using the [SF Salaries Dataset](https://www.kaggle.com/kaggle/sf-salaries) from Kaggle! Just follow along and complete the tasks outlined in bold below. The tasks will get harder and harder as you go along.

** Import pandas as pd.**

In [1]:
import pandas as pd
import numpy as np

** Read Salaries.csv as a dataframe called sal.**

In [4]:
sal = pd.read_csv('Salaries.csv')

In [6]:
sal.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

** Check the head of the DataFrame. **

** Use the .info() method to find out how many entries there are.**

In [7]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
Id                  148654 non-null int64
EmployeeName        148654 non-null object
JobTitle            148654 non-null object
BasePay             148045 non-null float64
OvertimePay         148650 non-null float64
OtherPay            148650 non-null float64
Benefits            112491 non-null float64
TotalPay            148654 non-null float64
TotalPayBenefits    148654 non-null float64
Year                148654 non-null int64
Notes               0 non-null float64
Agency              148654 non-null object
Status              0 non-null float64
dtypes: float64(8), int64(2), object(3)
memory usage: 13.0+ MB


**What is the average BasePay ?**

In [11]:
sal["BasePay"].mean()

66325.4488404877

** What is the highest amount of OvertimePay in the dataset ? **

In [12]:
sal["OvertimePay"].mean()

245131.88

** What is the job title of  JOSEPH DRISCOLL ? Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll). **

In [14]:
sal[sal['EmployeeName']=='JOSEPH DRISCOLL']['JobTitle']

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

** How much does JOSEPH DRISCOLL make (including benefits)? **

In [19]:
joseph = sal[sal['EmployeeName']=='JOSEPH DRISCOLL']
joseph[['BasePay','OvertimePay','OtherPay','Benefits']].sum(axis=1)
# expected:  270324.91

24    270324.91
dtype: float64

** What is the name of highest paid person (including benefits)?**

In [31]:
sal.iloc[sal['TotalPayBenefits'].idxmax()]

# Expected: NATHANIEL FORD

Id                                                               1
EmployeeName                                        NATHANIEL FORD
JobTitle            GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
BasePay                                                     167411
OvertimePay                                                      0
OtherPay                                                    400184
Benefits                                                       NaN
TotalPay                                                    567595
TotalPayBenefits                                            567595
Year                                                          2011
Notes                                                          NaN
Agency                                               San Francisco
Status                                                         NaN
Name: 0, dtype: object

** What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid?**

In [32]:
sal.iloc[sal['TotalPayBenefits'].idxmin()]

Id                                      148654
EmployeeName                         Joe Lopez
JobTitle            Counselor, Log Cabin Ranch
BasePay                                      0
OvertimePay                                  0
OtherPay                               -618.13
Benefits                                     0
TotalPay                               -618.13
TotalPayBenefits                       -618.13
Year                                      2014
Notes                                      NaN
Agency                           San Francisco
Status                                     NaN
Name: 148653, dtype: object

** What was the average (mean) BasePay of all employees per year? (2011-2014) ? **

In [34]:
sal.groupby('Year').mean()['BasePay']

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

Expected: 
Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

** How many unique job titles are there? **

In [36]:
sal['JobTitle'].nunique()

2159

** What are the top 5 most common jobs? **

In [45]:
sal['JobTitle'].value_counts().head()

## Or sal.groupby('JobTitle').count().sort_values(by='Id',ascending=False).head()

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

** How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?) **

In [53]:
sal[sal['Year'] == 2013]['JobTitle'].value_counts().where(lambda x:x == 1).dropna()

Mental Health Hearing Officer     1.0
Principal Museum Preparator       1.0
Special Assistant 21              1.0
Farmer                            1.0
Arborist Technician Supv II       1.0
Rent Board Supervisor             1.0
Assistant Industrial Hygienist    1.0
IS Operator-Journey               1.0
IS Trainer-Senior                 1.0
Cement Finisher Supervisor 2      1.0
Controller                        1.0
Director, Juvenile Hall           1.0
Field Svcs Asst Supv              1.0
Sr Employee Asst Counselor        1.0
Senior Psychiatric Orderly        1.0
Assistant Power House Operator    1.0
Senior Assessment Clerk           1.0
Special Assistant 16              1.0
Assistant Sheriff                 1.0
Senior Transit Traffic Checker    1.0
Statistician                      1.0
Victim & Witness Technician       1.0
Supervising Parts Storekeeper     1.0
Payroll Supervisor                1.0
Auto Body & Fender Wrk Sprv 1     1.0
Administrator, DPH                1.0
Real Estate 

** How many people have the word Chief in their job title? (This is pretty tricky) **

In [73]:
sal[sal['JobTitle'].str.contains('Chief')]

Id         EmployeeName                        JobTitle  \
36160    36161         Gregory Suhr                 Chief of Police   
36162    36163   Joanne Hayes-White          Chief, Fire Department   
36169    36170         Brendan Ward  Battlion Chief, Fire Suppressi   
36170    36171           Samson Lai  Battlion Chief, Fire Suppressi   
36171    36172     Rudy Castellanos  Battlion Chief, Fire Suppressi   
36172    36173           Marty Ross  Battlion Chief, Fire Suppressi   
36173    36174         Mark Kearney        Assistant Deputy Chief 2   
36179    36180       Michael Morris        Assistant Deputy Chief 2   
36181    36182         James Dudley                  Deputy Chief 3   
36190    36191        Thomas Harvey        Assistant Deputy Chief 2   
36196    36197          David Shinn                  Deputy Chief 3   
36197    36198        Kevin Cashman                  Deputy Chief 3   
36201    36202          Pete Fay Jr  Battlion Chief, Fire Suppressi   
36204    36205       Denise Schmitt                  Deputy Chief 3   
36208    36209         Kevin Taylor  Battlion Chief, Fire Suppressi   
36212    36213      Kirk Richardson  Battlion Chief, Fire Suppressi   
36215    36216         Lorrie Kalos  Battlion Chief, Fire Suppressi   
36220    36221      James Vannucchi  Battlion Chief, Fire Suppressi   
36222    36223     Bryan Rubenstein  Battlion Chief, Fire Suppressi   
36226    36227     Michael Thompson  Battlion Chief, Fire Suppressi   
36228    36229          Kevin Smith  Battlion Chief, Fire Suppressi   
36230    36231             Rex Hale  Battlion Chief, Fire Suppressi   
36233    36234            Jose Velo        Assistant Deputy Chief 2   
36241    36242        Jeffrey Myers    Emergency Medical Svcs Chief   
36247    36248          Lyn Tomioka                  Deputy Chief 3   
36248    36249        George Garcia  Battlion Chief, Fire Suppressi   
36250    36251       Michael Bryant  Battlion Chief, Fire Suppressi   
36256    36257         Michael Biel                  Deputy Chief 3   
36259    36260        Charles Crane  Battlion Chief, Fire Suppressi   
36263    36264         Mark Johnson  Battlion Chief, Fire Suppressi   
...        ...                  ...                             ...   
120012  120013         Mark A Naddy       Chief Stationary Engineer   
121045  121046      Rolando C Lopez       Chief Stationary Engineer   
122123  122124   Nerivaldo C Santos     Chief Fire Alarm Dispatcher   
122170  122171         James M Ryan                  Chief Surveyor   
122574  122575    Loretta M Cecconi                 Chief Dietitian   
123030  123031        Michael G Gay       Chief Stationary Engineer   
124548  124549         Paul H Mabry                  Chief Surveyor   
124647  124648       Mark Y Antonio                     Chief Clerk   
125497  125498      Artie L L Kelly                     Chief Clerk   
125525  125526      Tuti I Suardana                     Chief Clerk   
125670  125671        Sylvia N Thai                     Chief Clerk   
125673  125674      Deborah L Brown                     Chief Clerk   
125861  125862          Cuong Quach                     Chief Clerk   
125893  125894         Peter Aviles                     Chief Clerk   
125907  125908      Shahin P Shaikh                     Chief Clerk   
126041  126042    Donald P Margolis      Assistant Chief Attorney 1   
127965  127966    Sylvia M Selinger                     Chief Clerk   
128030  128031      Janice Gragasin                     Chief Clerk   
131266  131267     Cheryl L Coleman                     Chief Clerk   
131377  131378       Ann M Castillo        Chief Telephone Operator   
132799  132800    Judith A Boyajian      Assistant Chief Attorney 1   
133842  133843     Jennifer L Novak                     Chief Clerk   
137010  137011   Ryan L Butterfield                Chief Preparator   
137168  137169        Tony E Durkee                  Chief Surveyor   
137553  137554         Paula Jesson 

477

** Bonus: Is there a correlation between length of the Job Title string and Salary? **

In [85]:
sal['JobTitleLen'] = sal.apply(lambda row: len(row['JobTitle']), axis=1)
sal['JobTitleLen'].corr(sal['TotalPayBenefits'])

# Or the Code Below! Which is much faster!!

-0.036878445932606634

In [84]:
sal['JobTitleLen'] = sal['JobTitle'].apply(len)
sal[['JobTitleLen','TotalPayBenefits']].corr()

JobTitleLen  TotalPayBenefits
JobTitleLen          1.000000         -0.036878
TotalPayBenefits    -0.036878          1.000000

title_len  TotalPayBenefits
title_len          1.000000         -0.036878
TotalPayBenefits  -0.036878          1.000000

# Great Job!